File: Data Structures & Algorithms/binary-tree-diameter/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        /**
         * @param {TreeNode} tree
         * @param {number} dt
         */
        const f = (tree, acc) => {
            if (tree == null) return [acc, 0]
            const l = f(tree.left, acc)
            const r = f(tree.right, acc)
            const dl = tree.left ? (1 + l[1]) : 0
            const dr = tree.right ? (1 + r[1]) : 0
            console.log({ v: tree.val, l, r, dl, dr, acc })
            return [
                Math.max(acc, l[0], r[0], dl + dr),
                Math.max(dl, dr)
            ]
        }

        const res = f(root, 0)
        console.log(res)
        return res[0]
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import math


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def dfs(acc: int, node: Optional[TreeNode]) -> tuple[int, Optional[int]]:
            if node is None:
                return (acc, 0)

            (acc, left) = dfs(acc, node.left)
            (acc, right) = dfs(acc, node.right)

            value = max(left or 0, right or 0) + node.val
            print(acc, value)

            acc = max(acc, (left or 0) + (right or 0) + node.val, left or -1000, right or -1000)

            if value < 0:
                # Path を切る
                return (acc, None)
            else:
                # Path を追加
                return (acc, value)

        return dfs(-1000, root)[0]
